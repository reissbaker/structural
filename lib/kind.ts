import type { Comment, Either, DefaultIntersect, Validation } from "./type";
import type { TypeOf } from "./checks/type-of";
import type { InstanceOf } from "./checks/instance-of";
import type { Value } from "./checks/value";
import type { Arr } from "./checks/array";
import type { Dict, Struct, MergeIntersect } from "./checks/struct";
import type { MapType } from "./checks/map";
import type { SetType } from "./checks/set";
import type { Any } from "./checks/any";
import type { Is } from "./checks/is";
import type { Never } from "./checks/never";
import type { DeepPartial, PartialStruct } from "./checks/partial";

export type Kind = Any
                 | Never
                 | SetType<any>
                 | MapType<any, any>
                 | Dict<any>
                 | Struct<any>
                 | MergeIntersect<any, any, any, any>
                 | Arr<any>
                 | Value<any>
                 | InstanceOf<any>
                 | TypeOf<any>
                 | Either<any, any>
                 | DefaultIntersect<any, any>
                 | Validation<any>
                 | Is<any>
                 | Comment<any>
                 | PartialStruct<any>
                 | DeepPartial<any>
                 ;
