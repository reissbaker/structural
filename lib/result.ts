import { Type } from './type'
import { inspect } from 'util'

const MAX_ERR_LINE_LENGTH = 40
const INDENT = '  '

export const indent = (lines: string | string[], prefix = INDENT) =>
  (Array.isArray(lines) ? lines : lines.split("\n"))
    .map(line => prefix + line)
    .join("\n")

export const indentNext = (lines: string | string[], prefix = INDENT) => {
  const [first, ...rest] = Array.isArray(lines) ? lines : lines.split("\n")
  if (rest.length === 0) { return first }
  return [first, ...indent(rest, prefix).split("\n")].join("\n")
}

export function shouldWrap(msg: string | string[]): boolean {
  const parts = Array.isArray(msg) ? msg : [msg];

  const len = parts.map(s => s.length).reduce((m, x) => m + x, 0)
  if (len > MAX_ERR_LINE_LENGTH) { return true }

  const hasNewline = parts.map(s => s.includes("\n")).reduce((m, x) => m || x)
  return hasNewline
}

function quoteOrIndent(msg: string, suffix?: (inline: boolean) => string): string {
  if (shouldWrap(msg)) {
    return ("\n" +
      indent(msg) +
      (msg.endsWith("\n") ? '' : "\n") +
      (suffix ? suffix(false) : ''))
  }
  return '`' + msg + '`' + (suffix ? suffix(true) : '')
}

function subMessage(pre: string, msg: string): string {
  if (pre.includes("\n")) {
    return concat(pre, indentNext(msg))
  }
  return concat(pre, "\n" + indent(msg))
}

function concat(...msgs: string[]): string {
  const out = []
  for (let i = 0; i<msgs.length; i++) {
    const cur = msgs[i]
    const nxt = msgs[i+1]
    out.push(cur)
    if (!nxt || cur.endsWith("\n") ||  nxt.startsWith("\n")) {
      continue;
    }
    out.push(" ")
  }
  return out.join('')
}

type tostr = () => string

type ErrOpts<_> = {
    value: any,
    type: Type<any>,
    path?: PathElement[],
    causes?: Err<_>[]
}

const errSep = (inline: boolean) => inline ? ':' : 'because:'

export class Err<_> {
  path: PathElement[] // array defining access to the field in `data`. Will be [] if element is data
  value: any          // The invalid value
  type: Type<any>     // Expected type of value
  causes: Err<_>[]
  protected msg: string | tostr;

  static lift<_>(err: Err<_>, ...path: PathElement[]) {
    return new Err<_>(
      err.msg,
      {
        value: err.value,
        type: err.type,
        path: path.concat(err.path),
        causes: err.causes,
      }
    )
  }

  // TODO: disallow causes option
  static combine<_>(errs: Err<any>[], opts: ErrOpts<_>): Err<_> {
    return new Err(
      () => {
        if (errs.length === 1) { return errs[0].toString() }
        const errStrings = errs.map((e) => `- ${e}`).join("\n")
        return `failed multiple checks:\n${errStrings}`
      },
      { ...opts, causes: errs },
    )
  }

  constructor(msg: string | tostr, {path = [], value, type, causes = []}: ErrOpts<_>) {
    this.msg = msg
    this.path = path
    this.value = value
    this.type = type
    this.causes = causes
  }

  get message(): string {
    const msg = typeof this.msg === 'function' ?
      this.msg() : this.msg
    this.msg = msg
    return msg
  }

  protected rootCauses(parentPath: PathElement[]): Err<_>[] {
    if (this.causes.length === 0) {
      return [Err.lift(this, ...parentPath)]
    }
    return this.causes
      .map(c => c.rootCauses(parentPath.concat(this.path)))
      .reduce((x, m) => x.concat(m), [])
  }

  toError(): StructuralError {
    // force expanding all errors now so that the lifted errors in rootCauses
    // below are already memoized
    const asString = this.toStringWithValue()

    // flatten causes, which is more inspectable to downstream exception
    // handlers
    const causes = this.causes
      .map(c => c.rootCauses(this.path))
      .reduce((x, m) => x.concat(m), [])

    return new StructuralError(
      asString,
      {
        value: this.value,
        type: this.type,
        causes: causes,
      }
    );
  }

  toStringWithValue() {
    const pathPart = this.path.length ? ['at', `${pathToString(this.path)}:`] : []
    return concat(
      ...pathPart,
      'given value', quoteOrIndent(inspect(this.value)),
      'did not match expected type', subMessage(
        quoteOrIndent(this.type.toString(), errSep),
        this.message,
      )
    )
  }

  toStringNoValue() {
    return concat(
      'not of type',
      subMessage(
        quoteOrIndent(this.type.toString(), errSep),
        this.message,
      ),
    )
  }

  // Use this stringifier when embedding an error inside another error's message.
  // Avoids printing the value again for errors that donâ€™t modify the path
  toString() {
    if (this.path.length === 0) { return this.toStringNoValue() }
    return this.toStringWithValue()
  }
}

/**
 * KeyIndex represents a specific key in a map when the key itself is invalid.
 * Consider a map of Map<A, B>. If we assert for a concrete value that is actually
 * a Map<A | number, B>, we can use KeyIndex to indicate that it is the Nth key
 * which is invalid (because it is a number). This is even more useful if the key
 * is a struct, and some deep field of the struct has gone wrong.
 */
export class MapKeyIndex {
  readonly index: number
  constructor(n: number) {
    this.index = n
  }

  toString() {
    return `.keys()[${this.index}]`
  }

  apply(current: any) {
    const keys = Array.from(current.keys())
    if (keys.length <= this.index) {
      throw new Error(`${this} index out of bounds of keys ${keys.length} in ${current}`)
    }
    return keys[this.index]
  }
}

/**
 * For map types, this boxes the key so it's clear we're accessing a map. If we just used `any`
 * in the path, we couldn't be sure when pulling out KeyIndex values what those values meant.
 */
export class MapKey {
  readonly val: any
  constructor(v: any) {
    this.val = v
  }

  toString() {
    return `.get(${this.val.toString()})`
  }

  apply(current: any) {
    return current.get(this.val)
  }
}
type PathElement = string | number | MapKeyIndex | MapKey

// is there any point to this other than intellectual exercise??
export function lookupPath(val: any, path: PathElement[]): any {
  if (path.length === 0) {
    return val
  }

  let current = val
  path.forEach((key) => {
    if (key instanceof MapKeyIndex) {
      current = key.apply(current)
      return
    }
    if (key instanceof MapKey) {
      current = key.apply(current)
      return
    }
    current = current[key]
  })
  return current
}

export function pathToString(lookupPath: PathElement[]): string {
  let out: string[] = []
  for (let i = 0; i<lookupPath.length; i++) {
    const el = lookupPath[i]
    if (typeof el === 'string') {
      out.push(`.${el}`)
      continue
    }
    if (typeof el === 'number') {
      out.push(`[${el}]`)
      continue
    }
    out.push(el.toString())
  }
  return out.join('')
}

export class StructuralError extends Error {
  readonly value: any          // The invalid value
  readonly type: Type<any>     // Expected type of value
  readonly causes: Err<any>[]  // Flattened causes of this error

  constructor(msg: string, opts: Pick<StructuralError, 'value' | 'type' | 'causes'>) {
    super(msg)
    this.value = opts.value
    this.type = opts.type
    this.causes = opts.causes
  }
}

export type Result<T> = T | Err<T>;
