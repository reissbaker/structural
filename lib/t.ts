/*
 * This is a smaller import, `t`, that only contains the basic matchers and code that would either
 * necessarily be loaded to run them, or that takes 0 runtime space (e.g. type-only files).
 * It exists to make it easy to import everything you would typically use to define schemas, but
 * leave out optional stuff like converting to TypeScript, JSON Schema, etc from bundles, assuming
 * you use a tree-shaking compiler. This effectively means that adding more conversion utilities
 * doesn't bloat imports, and still provides reasonable import DX so that you don't need to have
 * individual import statements for every single type. Instead of the old style:
 *
 *     import * as t from "structural";
 *
 * You can instead do:
 *
 *     import { t } from "structural";
 *
 * And save a little bundle space.
 */

export * from "./result";
export * from "./type";
export * from "./get-type";

export * from "./checks/type-of";
export * from "./checks/instance-of";
export * from "./checks/value";
export * from "./checks/array";
export * from "./checks/struct";
export * from "./checks/map";
export * from "./checks/set";
export * from "./checks/primitives";
export * from "./checks/any";
export * from "./checks/is";
export * from "./checks/never";

export * from "./kind";
