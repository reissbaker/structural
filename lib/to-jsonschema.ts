import { Type, Comment, Either, Intersect, Validation } from "./type";
import { TypeOf } from "./checks/type-of";
import { InstanceOf } from "./checks/instance-of";
import { Value } from "./checks/value";
import { Arr } from "./checks/array";
import { Struct, MissingKey, OptionalKey } from "./checks/struct";
import { Dict } from "./checks/dict";
import { MapType } from "./checks/map";
import { SetType } from "./checks/set";
import { Any } from "./checks/any";
import { Is } from "./checks/is";
import { Never } from "./checks/never";
import { PartialStruct, DeepPartial } from "./checks/partial";
import { Kind } from "./kind";

export const JSON_SCHEMA_VERSION = "https://json-schema.org/draft/2020-12/schema" as const;
type TopLevel<T> = T & {
  $schema: typeof JSON_SCHEMA_VERSION,
  title: string,
};

// TODO: a title annotation built into structural. description can be transformed from comments.
// you should also have a variant of toJSONSchema that mandates top-level titles by insisting that
// the type passed in is a Title type instead of a Kind type.
type Annotated<T> = T & {
  description?: string,
};

export type JSONValue = null
                      | string
                      | number
                      | boolean
                      | Array<JSONValue>
                      | { [key: string]: JSONValue }
                      ;

export type JSONSchema = Annotated<{ type: "string" }>
                       | Annotated<{ type: "number" }>
                       | Annotated<{ type: "boolean" }>
                       | Annotated<{ type: "null" }>
                       | Annotated<{}> // {} is JSON schema's any type
                       | Annotated<{
                         type: "object",
                         required: string[],
                         properties: {
                           [key: string]: JSONSchema,
                         },
                       }>
                       | Annotated<{
                         type: "object",
                         // Dict<T> type is { properties: {}, additionalProperties: T }
                         properties: {},
                         additionalProperties?: JSONSchema,
                       }>
                       | Annotated<{
                         type: "array",
                         items?: JSONSchema,
                       }>
                       | Annotated<{
                         // intersection
                         allOf: JSONSchema[],
                       }>
                       | Annotated<{
                         // union
                         anyOf: JSONSchema[],
                       }>
                       | Annotated<{
                         // technically this can be represented as anyOf([ ...consts ]), but for
                         // readability if all of a union is const values imo you should output an
                         // enum rather than the more complex type
                         enum: Array<JSONValue>
                       }>
                       | Annotated<{
                         const: JSONValue,
                       }>
                       ;

type Options = {
  errorOnValidations?: boolean,
  errorOnIs?: boolean,
  errorOnNever?: boolean,
};
const defaultOpts: Required<Options> = {
  errorOnValidations: true,
  errorOnIs: true,
  errorOnNever: true,
};

export function toJSONSchema(title: string, type: Kind, opts?: Options): TopLevel<JSONSchema> {
  const options = {
    ...defaultOpts,
    ...opts,
  };

  return addMetadata(title, typeToSchema(type, options));
}

function typeToSchema(type: Kind, options: Required<Options>): JSONSchema {
  if(type instanceof Comment) {
    return {
      description: formatCommentString(type.commentStr),
      ...typeToSchema(type.wrapped, options)
    };
  }
  if(type instanceof Either) return fromEither(type, options);
  if(type instanceof Intersect) return fromIntersect(type, options);
  if(type instanceof Validation) return fromValidation(type, options);
  if(type instanceof TypeOf) return fromTypeof(type);
  if(type instanceof InstanceOf) {
    throw `Structural instanceOf types can't be converted to JSON Schema`;
  }
  if(type instanceof Value) return fromValue(type);
  if(type instanceof Arr) return fromArray(type, options);
  if(type instanceof Struct) return fromStruct(type, options);
  if(type instanceof Dict) return fromDict(type, options);
  if(type instanceof MapType) {
    throw `Structural Map types can't be converted to JSON Schema; consider using a dict`;
  }
  if(type instanceof SetType) {
    throw `Structural Set types can't be converted to JSON Schema; consider using an array`;
  }
  if(type instanceof Any) {
    return {};
  }
  if(type instanceof Is) return fromIs(type, options);
  if(type instanceof Never) return fromNever(type, options);
  if(type instanceof DeepPartial) return fromDeepPartial(type, options);

  return fromPartial(type, options);
}

function areSerializableValues(types: Array<Type<any>>): types is Array<Value<JSONValue>> {
  for(const type of types) {
    if(!(type instanceof Value)) return false;
    if(!isSerializable(type.val)) return false;
  }
  return true;
}

function isSerializable(value: any): value is JSONValue {
  try {
    JSON.parse(JSON.stringify(value));
  } catch {
    return false;
  }
  return true;
}

function fromEither(type: Either<any, any>, options: Required<Options>): JSONSchema {
  // While the Either type could theoretically just map to recursive anyOf schemas, in practice it's
  // much more readable to have it attempt to generate an enum list of all values (if it's a bunch
  // of values unioned together, AND the values are all serializable to JSON), or a flat anyOf list
  // otherwise. So, we walk the tree and collect all the types, and do that.
  const allTypes = flatTypes(Either, type);

  if(areSerializableValues(allTypes)) {
    return {
      enum: allTypes.map(t => t.val),
    };
  }

  return {
    anyOf: allTypes.map(type => typeToSchema(type, options)),
  };
}

function fromIntersect(type: Intersect<any, any>, options: Required<Options>): JSONSchema {
  if(type.r instanceof Validation) {
    if(options.errorOnValidations) {
      throw `Structural type contains a validation, but errorOnValidations was set to true`;
    }
    return {
      description: type.r.desc,
      ...typeToSchema(type.l, options),
    };
  }
  return {
    allOf: flatTypes(Intersect, type).map(t => typeToSchema(t, options)),
  };
}

function fromValidation(type: Validation<any>, options: Required<Options>): JSONSchema {
  if(options.errorOnValidations) {
    throw `Structural type contains a validation, but errorOnValidations was set to true`;
  }

  return {
    description: type.desc
  };
}

function fromTypeof(type: TypeOf<any>): JSONSchema {
  switch(type.typestring) {
    case "string": return { type: "string" };
    case "number": return { type: "number" };
    case "boolean": return { type: "boolean" };
    case "object": return { type: "object", properties: {} };
  }
  throw `Can't convert Structural TypeOf<${type.typestring}> checks into JSON Schema`
}

function fromValue(type: Value<any>): JSONSchema {
  if(!isSerializable(type.val)) {
    throw `Value type expects ${type.val}, but it isn't serializable to JSON`;
  }

  if(type.val === null) return { type: "null" };

  return {
    const: type.val,
  };
}

function fromArray(type: Arr<any>, options: Required<Options>): JSONSchema {
  return {
    type: 'array',
    items: typeToSchema(type.elementType, options),
  };
}

function fromStruct(type: Struct<any>, options: Required<Options>): JSONSchema {
  const properties: { [key: string]: JSONSchema } = {};
  const required: string[] = [];
  const keys = Object.keys(type.definition);
  for(const key of keys) {
    const val = type.definition[key];
    if(val instanceof MissingKey || val instanceof OptionalKey) {
      properties[key] = typeToSchema(val.type, options);
    }
    else {
      required.push(key);
      properties[key] = typeToSchema(val, options);
    }
  }

  return {
    type: 'object',
    required, properties,
  };
}

function fromDict(type: Dict<any>, options: Required<Options>): JSONSchema {
  return {
    type: 'object',
    properties: {},
    additionalProperties: typeToSchema(type.valueType, options),
  };
}

function fromIs(type: Is<any>, options: Required<Options>) {
  if(options.errorOnIs) {
    throw `Structural type contains an Is check, but errorOnIs was set to true`;
  }

  return {
    description: type.name,
  };
}

function fromNever(_: Never, options: Required<Options>): JSONSchema {
  if(options.errorOnNever) {
    throw `Structural type contains a Never check, but errorOnNever was set to true`;
  }

  return {
    allOf: [ { type: "string" }, { type: "number" } ],
  };
}

function fromDeepPartial(type: DeepPartial<any>, options: Required<Options>): JSONSchema {
  // DeepPartial's .struct accessor is actually a rewritten struct with all keys made deeply
  // optional
  return fromStruct(type.struct, options);
}

function fromPartial(type: PartialStruct<any>, options: Required<Options>): JSONSchema {
  // Since Partial<T> only makes top-level keys optional, we just modify the required array
  const schema = fromStruct(type.struct, options);
  if('required' in schema) schema.required = [];
  return schema;
}

function flatTypes(
  klass: { new(...args: any): Either<any, any> | Intersect<any, any> },
  node: Type<any>
): Array<Type<any>> {
  if(node instanceof klass) {
    return flatTypes(klass, node.l).concat(flatTypes(klass, node.r));
  }
  return [ node ];
}

function formatCommentString(commentStr: string) {
  const commentLines = commentStr.split("\n").map(line => {
    return line.trim();
  }).filter(line => line !== "");
  return commentLines.join("\n");
}

function addMetadata(title: string, schema: JSONSchema): TopLevel<JSONSchema> {
  return {
    $schema: JSON_SCHEMA_VERSION,
    title,
    ...schema,
  };
}
