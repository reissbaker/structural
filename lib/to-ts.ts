import { Type, Comment, Either, Intersect, Validation } from "./type";
import { TypeOf } from "./checks/type-of";
import { InstanceOf } from "./checks/instance-of";
import { Value } from "./checks/value";
import { Arr } from "./checks/array";
import { Struct, OptionalKey } from "./checks/struct";
import { Dict } from "./checks/dict";
import { MapType } from "./checks/map";
import { SetType } from "./checks/set";
import { Any } from "./checks/any";
import { Is } from "./checks/is";
import { Never } from "./checks/never";
import { Kind } from "./kind";

type ToTypescriptOpts = {
  useReference?: {
    [ref: string]: Type<any>,
  },

  indent: string,
  indentLevel: number,
};

export type TypescriptUserOpts = Partial<ToTypescriptOpts> & {
  assignToType?: string,
};

type SingleConversionWithOpts = [ type: Kind, userOpts: TypescriptUserOpts ];
type SingleConversion = [ type: Kind ];
type MultipleConversion = [ types: { [name: string]: Kind } ];

export function toTypescript(...args: SingleConversion | SingleConversionWithOpts | MultipleConversion): string {
  if(args.length === 2) {
    const [ type, userOpts ] = args;
    const opts = Object.assign({ indent: "  ", indentLevel: 0 }, userOpts);
    // assignToType is only valid at the top level, so delete it if it exists
    delete opts.assignToType;

    const ts = toTS(type, opts);

    if(userOpts.assignToType) return `type ${userOpts.assignToType} = ${ts};`;
    return ts;
  }

  const arg = args[0]
  if(arg instanceof Type) return toTypescript(arg, {});

  const keys = Object.keys(arg);
  if(keys.length === 0) return "";
  const output: string[] = [];
  for(const key of keys) {
    const refs = Object.assign({}, arg);
    delete refs[key];
    output.push(toTypescript(arg[key], {
      useReference: refs,
      assignToType: key,
    }));
  }
  return output.join("\n\n");
}

function toTS(type: Kind, opts: ToTypescriptOpts): string {
  if(opts.useReference) {
    for(const key in opts.useReference) {
      const val = opts.useReference[key];
      if(val === type) return key;
    }
  }

  if(type instanceof Comment) return fromComment(type, opts);
  if(type instanceof Either) return fromEither(type, opts);
  if(type instanceof Intersect) return fromIntersect(type, opts);
  if(type instanceof Validation) return fromValidation();
  if(type instanceof TypeOf) return fromTypeof(type);
  if(type instanceof InstanceOf) return fromInstanceOf(type);
  if(type instanceof Value) return fromValue(type);
  if(type instanceof Arr) return fromArr(type, opts);
  if(type instanceof Struct) return fromStruct(type, opts);
  if(type instanceof Dict) return fromDict(type, opts);
  if(type instanceof MapType) return fromMap(type, opts);
  if(type instanceof SetType) return fromSet(type, opts);
  if(type instanceof Any) return fromAny();
  if(type instanceof Is) return fromIs(type);
  return fromNever(type);
}

function fromComment(c: Comment<any>, opts: ToTypescriptOpts) {
  const i = indent(opts);
  const commentLines = formatCommentString(c.commentStr, opts);
  return `${commentLines}\n${i}${toTS(c.wrapped, opts)}`;
}

function formatCommentString(commentStr: string, opts: ToTypescriptOpts) {
  const i = indent(opts);

  const commentLines = commentStr.split("\n").map(line => {
    return line.trim();
  }).filter(line => line !== "");

  if(commentLines.length === 0) return "";
  if(commentLines.length === 1) {
    return `// ${commentLines[0]}`;
  }

  const lines = [ '/*' ]
  for(const line of commentLines) {
    lines.push(`${i} * ${line.trim()}`);
  }
  lines.push(`${i} */`);
  return lines.join("\n");
}

// TODO: Should either strip immediate-child comments?
function fromEither(e: Either<any, any>, opts: ToTypescriptOpts) {
  const i = indent(opts);
  return [
    toTS(e.l, opts),
    `${i}${opts.indent}| ${toTS(e.r, {...opts, indentLevel: opts.indentLevel + 1})}`,
  ].join("\n");
}

// TODO: Should intersect strip immediate-child comments?
function fromIntersect(i: Intersect<any, any>, opts: ToTypescriptOpts) {
  // Handle validations chained with actual TS types, converting them to comments
  if(i.left instanceof Validation) return toTS(new Comment(i.left.desc, i.r), opts);
  if(i.r instanceof Validation) return toTS(new Comment(i.r.desc, i.left), opts);

  const indentation = indent(opts);
  return [
    toTS(i.left, opts),
    `${indentation}${opts.indent}& ${toTS(i.r, {...opts, indentLevel: opts.indentLevel + 1})}`,
  ].join("\n");
}

function fromValidation(): string {
  throw new Error(
    "Can't convert arbitrary validation functions to TypeScript types; make sure to .and() it " +
      "with a valid type, and it will be converted into a comment above the type"
  );
}

function fromTypeof(t: TypeOf<any>): string {
  switch(t.typestring) {
    case "undefined": return t.typestring;
    case "object": return "Object";
    case "boolean": return t.typestring;
    case "number": return t.typestring;
    case "bigint": return "BigInt";
    case "string": return t.typestring;
    case "symbol": return "Symbol";
    case "function": return "Function";
  }
}

function fromInstanceOf(i: InstanceOf<any>) {
  if(!i.klass.name) throw new Error("Can't convert anonymous classes to TypeScript");
  return `${i.klass.name}`;
}

function fromValue(v: Value<any>) {
  const vType = typeof v.val;
  if(vType !== "string" && vType !== "number" && vType !== "boolean" && v.val !== null && v.val !== undefined) {
    throw new Error(
      "Only string, numeric, undefined, boolean, and null value types can be auto-converted to TypeScript"
    );
  }
  if(vType === "string") return JSON.stringify(v.val);

  return `${v.val}`;
}

function fromArr(a: Arr<any>, opts: ToTypescriptOpts) {
  return `Array<${toTS(a.elementType, opts)}>`;
}

function fromStruct(s: Struct<any>, opts: ToTypescriptOpts) {
  const lines = [ "{" ];
  const keyOpts = {
    ...opts,
    indentLevel: opts.indentLevel + 1,
  };
  const keyIndent = indent(keyOpts);
  const keys = Object.keys(s.definition);

  for(let i = 0; i < keys.length; i++) {
    const key = keys[i];
    const keyType = [ key ];
    const val = s.definition[key];
    if(val instanceof OptionalKey) keyType.push("?");
    keyType.push(": ");
    const stripped = stripOuterComments(val);
    if(stripped.comments.length > 0) {
      // Visually separate the start of a commented field unless it's the first field
      if(i !== 0) lines.push("");
      // Put the comment on the line above the key
      lines.push(keyIndent + formatCommentString(stripped.comments.join("\n"), keyOpts));
    }
    keyType.push(toTS(stripped.inner, keyOpts));
    keyType.push(",");
    lines.push(keyIndent + keyType.join(""));
    // Visually separate the end of a commented field, unless it's the last field
    if(stripped.comments.length > 0 && i !== keys.length - 1) lines.push("");
  }
  lines.push(indent(opts) + "}");

  return lines.join("\n");
}

type StrippedComments = {
  comments: string[],
  inner: Kind,
};

function stripOuterComments(t: Kind | OptionalKey<any>): StrippedComments {
  if(t instanceof OptionalKey) return stripOuterComments(t.type);
  if(t instanceof Comment) {
    const inner = stripOuterComments(t.wrapped);
    return {
      comments: [ t.commentStr, ...inner.comments ],
      inner: inner.inner,
    }
  }

  if(t instanceof Intersect) {
    if(t.left instanceof Validation) return stripOuterComments(new Comment(t.left.desc, t.r));
    if(t.r instanceof Validation) return stripOuterComments(new Comment(t.r.desc, t.left));
  }

  return {
    comments: [],
    inner: t,
  };
}

function fromDict(d: Dict<any>, opts: ToTypescriptOpts) {
  const i = indent(opts);
  const valString = toTS(d.valueType, {...opts, indentLevel: opts.indentLevel + 1});

  // For single-line values, return a single-line dict
  if(valString.indexOf("\n") < 0) return `{[${d.namedKey}: string]: ${valString}}`;

  // For multiline values, return a multiline dict
  return [
    "{",
    `${i}${opts.indent}[key: string]: ${valString}`,
    `${i}}`,
  ].join("\n");
}

function fromMap(m: MapType<any, any>, opts: ToTypescriptOpts) {
  const keyString = toTS(m.keyType, opts);
  const valString = toTS(m.valueType, opts);
  return `Map<${keyString}, ${valString}>`;
}

function fromSet(s: SetType<any>, opts: ToTypescriptOpts) {
  const valString = toTS(s.valueType, opts);
  return `Set<${valString}>`;
}

function fromAny() {
  return "any";
}

function fromIs(i: Is<any>) {
  return i.name;
}

// Despite not using the type, we take it to help ensure exhaustiveness checking in the toTS fn
function fromNever(_: Never) {
  return "never";
}

function indent(opts: ToTypescriptOpts) {
  return opts.indent.repeat(opts.indentLevel);
}
