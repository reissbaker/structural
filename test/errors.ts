/* noStripWhitespace - disable vim stripping trailing whitespace on file write.
* This is important for the expected error message strings below */
import * as t from '..'
import { indent, indentNext } from '../lib/result'

describe("error integration tests", () => {
  const User = t.subtype({
    username: t.str,
    age: t.optional(t.num),
    email: t.str.validate('has @ symbol', (val) => !!val.match(/@/))
  })

  const Quote = t.exact({
    text: t.str,
    author: t.str,
    date: t.optional(t.instanceOf(Date)),
  })

  const Pic = t.subtype({
    uri: t.str.validate('starts with http', (val) => !!val.match(/^http/)),
    alt: t.str,
  })

  const Post = t.subtype({
    owner: User,
    contents: t.array(Quote.or(Pic)),
  })

  function value(): t.GetType<typeof Post> {
    return {
      owner: {
        username: 'the duke',
        age: 42,
        email: 'duke.nuke@example.com',
      },
      contents: [
        {
          text: 'Those who do not study history are doomed to repeat it',
          author: 'Albert Einstien',
          date: new Date(2001),
        },
        {
          uri: 'https://example.com/pics/albert.jpg',
          alt: 'Mr. Doctor Albert Einstien wearing a tweed jacket',
        }
      ]
    }
  }

  function captureErr(block: () => void): Error {
    let err: Error
    try {
      block()
    } catch (e) {
      err = e
      return err
    }
    throw "no error produced"
  }

  test('the default value is valid', () => {
    Post.assert(value())
  })

  test("either enumerates the missed types", () => {
    const val = value();
    (val.contents[1] as any).uri = 'example.com'

    expect(() => {
      Post.assert(val)
    }).toThrowErrorMatchingSnapshot()
  })

  test("exact prints fields nice", () => {
    const val = value().contents[0]
    ;(val as any).badBoy = true

    expect(() => {
      Quote.assert(val)
    }).toThrowErrorMatchingSnapshot()
  })

  test("error contains a full .causes", () => {
    const err = captureErr(() => {
      const val = value();
      (val.contents[1] as any).uri = 'example.com'
      Post.assert(val)
    })

    expect(err).toBeInstanceOf(t.StructuralError)
    const se = err as t.StructuralError
    expect(se.causes).toHaveLength(3)
    expect(se.causes[0].toString()).toMatch(/at .contents\[1].text/)
    expect(se.causes[1].toString()).toMatch(/at .contents\[1].author/)
    expect(se.causes[2].toString()).toMatch(/at .contents\[1].uri/)
  })

  describe("helpers", () => {
    describe("indentNext", () => {
      test("does not make any changes to single-line strings", () => {
        expect(indentNext("foo")).toEqual("foo")
      })

      test("indents lines after the first line", () => {
        expect(indentNext("foo\nbar\nbaz")).toEqual("foo\n  bar\n  baz")
      })
    })

    describe("indent", () => {
      test("indents all lines", () => {
        expect(indent("foo\nbar")).toEqual("  foo\n  bar")
      })
    })
  })
})
