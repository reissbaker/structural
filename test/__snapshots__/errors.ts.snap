// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error integration tests either enumerates the missed types 1`] = `
"given value
  { owner:
     { username: 'the duke', age: 42, email: 'duke.nuke@example.com' },
    contents:
     [ { text: 'Those who do not study history are doomed to repeat it',
         author: 'Albert Einstien',
         date: 1970-01-01T00:00:02.001Z },
       { uri: 'example.com',
         alt: 'Mr. Doctor Albert Einstien wearing a tweed jacket' } ] }
did not match expected type
  { owner: { username: string,
      age?: number,
      email: string & validate(has @ symbol, (val) => !!val.match(/@/)) },
    contents: Array<{ text: string, author: string, date?: Date } |
    { uri: string & validate(starts with http, (val) => !!val.match(/^http/)),
      alt: string }> }
because: at .contents[1]: given value
    { uri: 'example.com',
      alt: 'Mr. Doctor Albert Einstien wearing a tweed jacket' }
  did not match expected type
    { text: string, author: string, date?: Date } |
    { uri: string & validate(starts with http, (val) => !!val.match(/^http/)),
      alt: string }
  because: matched none of 2 types:
    | { text: string, author: string, date?: Date }
      failed multiple checks:
      - at .text: given value \`undefined\` did not match expected type \`string\`:
        key missing
      - at .author: given value \`undefined\` did not match expected type \`string\`:
        key missing
    | { uri: string & validate(starts with http, (val) => !!val.match(/^http/)),
        alt: string }
      at .uri: given value \`'example.com'\` did not match expected type
        validate(starts with http, (val) => !!val.match(/^http/))
      because: validation returned false"
`;

exports[`error integration tests exact prints fields nice 1`] = `
"given value
  { text: 'Those who do not study history are doomed to repeat it',
    author: 'Albert Einstien',
    date: 1970-01-01T00:00:02.001Z,
    badBoy: true }
did not match expected type
  { text: string, author: string, date?: Date }
because: at .badBoy: given value \`true\` did not match expected type \`never\`:
    unknown key \`badBoy\` should not exist"
`;
